nixos/modules/services/hardware/usbmuxd.nix

```diff
@@ -0,0 +1,25 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+{
+  options.services.usbmuxd.enable = mkOption {
+    type = types.bool;
+    default = false;
+    description = ''
+      Enable the usbmuxd ("USB multiplexing daemon") service. This daemon is in
+      charge of multiplexing connections over USB to an iOS device. This is
+      needed for transferring data from and to iOS devices (see ifuse). Also
+      this may enable plug-n-play tethering for iPhones.
+    '';
+  };
+
+  config = mkIf config.services.usbmuxd.enable {
+    systemd.services.usbmuxd = {
+      description = "usbmuxd";
```

Comment:
Does this need to run as root? Have you tested your system with this module enabled?

Reply:
I have tested this module, using it on my system (with unstable).

I tried to make it work with non-root, that is by having it being started by non-root, which gave me the problem that it wants to create `/var/run/usbmuxd.pid` (even when run in foreground), which isn't possible as non-root afaik. It complains if the file exists already and the location is hardcoded, would require patching to change it.

Now what I haven't tried is using its `-U <user>` which changes to the given user after startup, but since it still has root permissions at start anyways and the additional effort to create a user I decided it's not worth it. On second thought, `-U` would be a good idea, since the real danger of it isn't in usbmuxd itself, but in the devices you connect, which only happens after startup. I'll have a look at it again later today.

Note: User needs to be in `disk` group.

Reply:
We could actually make it work. 3 things are needed:

1) Add RuntimeDirectory = "usbmuxd" to the service definition

2) Patch src/main.c to change the path to the pid from ```/var/run/usbmuxd.pid``` to ```/run/usbmuxd/usbmuxd.pid```.

3) Patch src/main.c to change the path to the socket.

Reply:
Is it really worth adding a patch (and I'd think this shouldn't be applied to the package itself) just so the service can run as non-root for startup though? It already switches to the given user after startup with the -U option.

Reply:
I'd argue that it's in our interest to make NixOS as safe as possible (within reason obviously) even if that means patching upstream to act more sane. Especially since the patching is just 2 sed invocations.